homeassistant/components/openai_conversation/__init__.py

```diff
@@ -187,41 +235,88 @@ async def async_process(
-            )
+        tool_calls = True
+        while tool_calls:
+            try:
+                result = await client.chat.completions.create(
+                    model=model,
+                    messages=messages,
+                    tools=[x[0] for x in self.tools],
+                    max_tokens=max_tokens,
+                    top_p=top_p,
+                    temperature=temperature,
+                    user=conversation_id,
+                )
+            except openai.OpenAIError as err:
+                intent_response = intent.IntentResponse(language=user_input.language)
+                intent_response.async_set_error(
+                    intent.IntentResponseErrorCode.UNKNOWN,
+                    f"Sorry, I had a problem talking to OpenAI: {err}",
+                )
+                return conversation.ConversationResult(
+                    response=intent_response, conversation_id=conversation_id
+                )
+
+            _LOGGER.debug("Response %s", result)
+            response = result.choices[0].message
+            messages.append(response)
+            tool_calls = response.tool_calls
+
+            if tool_calls:
+                for tool_call in tool_calls:  # type: ignore[attr-defined]
+                    function_response = await self.async_call_tool(
+                        user_input.context,
+                        tool_call.function.name,
+                        tool_call.function.arguments,
+                    )
+                    messages.append(
+                        {
+                            "role": "tool",
+                            "tool_call_id": tool_call.id,
+                            "name": tool_call.function.name,
+                            "content": function_response,
+                        }
+                    )
 
-        _LOGGER.debug("Response %s", result)
-        response = result.choices[0].message.model_dump(include={"role", "content"})
-        messages.append(response)
         self.history[conversation_id] = messages
 
         intent_response = intent.IntentResponse(language=user_input.language)
-        intent_response.async_set_speech(response["content"])
+        intent_response.async_set_speech(response.content)
         return conversation.ConversationResult(
             response=intent_response, conversation_id=conversation_id
         )
 
-    def _async_generate_prompt(self, raw_prompt: str) -> str:
+    def _async_generate_prompt(
+        self, raw_prompt: str, device_id: str | None, user_id: str | None
+    ) -> str:
         """Generate a prompt for the user."""
         return template.Template(raw_prompt, self.hass).async_render(
             {
                 "ha_name": self.hass.config.location_name,
+                "device_id": device_id,
+                "user_id": user_id,
             },
             parse_result=False,
         )
+
+
+@callback
+async def async_register_tool(
```

Comment:
I don't think that this is the right approach. 

We should generate a single, central, API for LLMs that we can then make available to all LLMs that support APIs. And we need to be in control of what exact functions are available.

Reply:
Ok, let me think about it.

Reply:
Are you on Discord? Hit me up `@balloob`